bindings/rust/extended/s2n-tls/src/callbacks/cert_validation.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use s2n_tls_sys::*;
+
+use crate::{
+    connection::Connection,
+    error::{Error, Fallible},
+};
+use std::ptr::NonNull;
+
+pub struct CertValidationInfo {
+    info: NonNull<s2n_cert_validation_info>,
+}
+
+impl CertValidationInfo {
+    pub(crate) fn from_ptr(info: *mut s2n_cert_validation_info) -> Result<Self, Error> {
+        let info = NonNull::new(info).ok_or(Error::INVALID_INPUT)?;
+        Ok(CertValidationInfo { info })
+    }
+
+    pub(crate) fn as_ptr(&self) -> *mut s2n_cert_validation_info {
+        self.info.as_ptr()
+    }
+
+    /// Corresponds to [s2n_cert_validation_accept].
+    pub fn accept(&self) -> Result<(), Error> {
+        unsafe { s2n_cert_validation_accept(self.as_ptr()).into_result() }?;
+        Ok(())
+    }
+
+    /// Corresponds to [s2n_cert_validation_reject].
+    pub fn reject(&self) -> Result<(), Error> {
+        unsafe { s2n_cert_validation_reject(self.as_ptr()).into_result() }?;
+        Ok(())
+    }
+}
+
+pub trait CertValidationCallbackSync: 'static + Send + Sync {
+    fn handle_validation(
+        &self,
+        connection: &mut Connection,
+        validation_info: &CertValidationInfo,
+    ) -> Result<(), Box<dyn std::error::Error>>;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{connection::Connection, security, testing::*};
+
+    type Error = Box<dyn std::error::Error>;
+
+    struct ValidationContext {
+        accept: bool,
+    }
+
+    struct SyncCallback(Counter);
+    impl CertValidationCallbackSync for SyncCallback {
+        fn handle_validation(
+            &self,
+            conn: &mut Connection,
+            info: &CertValidationInfo,
+        ) -> Result<(), Error> {
+            self.0.increment();
+            let context = conn.application_context::<ValidationContext>().unwrap();
+
+            match context.accept {
+                true => info.accept()?,
+                false => info.reject()?,
+            }
+            Ok(())
```

Comment:
I wonder if we could make this a bit nicer to use in Rust? Like, could we have the callback return a bool for accept/reject, and then call the actual accept/reject APIs on the info struct for the user maybe? The current way of calling the info struct seems a little strange for Rust.

Maybe it's fine as is though, especially if this API is somewhat temporary anyway.